command/debug.go

```
@@ -524,140 +499,225 @@ func (c *DebugCommand) captureStaticTargets(client *api.Client) error {
 
 // capturePollingTargets captures all dynamic targets over the specified
 // duration and interval.
-func (c *DebugCommand) capturePollingTargets(client *api.Client) error {
-	startTime := time.Now()
-	durationCh := time.After(c.flagDuration + debugDurationGrace)
-
-	// Track current interval count to show progress.
-	var idxCount, mIdxCount int
-
-	// errCh is used to capture any polling errors and recorded in the index
-	errCh := make(chan *captureError)
+func (c *DebugCommand) capturePollingTargets() error {
+	var g run.Group
 
-	var serverStatusCollection []map[string]interface{}
-	var replicationStatusCollection []map[string]interface{}
-	var metricsCollection []map[string]interface{}
-	var hostInfoCollection []map[string]interface{}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), c.flagDuration+debugDurationGrace)
 
-	// Start a goroutine to collect data as soon as it's received
-	collectCh := make(chan struct{})
-	go func() {
+	// This run group watches for interrupt or duration
+	g.Add(func() error {
 		for {
 			select {
-			case metricsEntry := <-c.metricsCh:
-				metricsCollection = append(metricsCollection, metricsEntry)
-			case serverStatusEntry := <-c.serverStatusCh:
-				serverStatusCollection = append(serverStatusCollection, serverStatusEntry)
-			case replicationStatusEntry := <-c.replicationStatusCh:
-				replicationStatusCollection = append(replicationStatusCollection, replicationStatusEntry)
-			case hostInfoEntry := <-c.hostInfoCh:
-				hostInfoCollection = append(hostInfoCollection, hostInfoEntry)
-			case <-c.doneCh:
-				close(collectCh)
-				return
+			case <-c.ShutdownCh:
+				return nil
```

These should return an error instead of nil, to trigger the rungroup ending.

---

My mistake, thinking of errgroup.